Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,322 @@
+   hP : ∃ i j : I , i ≠ j
+   /-- For any participant `i`, there exists another participant `j` who is different from `i`. -/
+   hP' :  ∀ i : I , ∃ j, i ≠  j := two_different_elements hP
+   /-- A function mapping each participant to their valuation. -/
+   v : I → ℝ
+
+namespace Auction
+
+variable {a : Auction} (b : a.I → ℝ )
+
+instance : Fintype a.I := a.hF
+
+
+/-!###  Helper Functions and Definitions -/
+
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxb : ℝ  := Finset.sup' Finset.univ (⟨ a.hI.default ,  (Finset.mem_univ _)⟩ ) b
+
+/-- There exists a participant `i` whose bid equals the highest bid. -/
+lemma exists_max : ∃ i: a.I, b i = a.maxb b := by
+   obtain ⟨  i , _ ,h2⟩ := Finset.exists_mem_eq_sup' (⟨ a.hI.default, (Finset.mem_univ _)⟩ ) b
+   exact ⟨ i, symm h2⟩
+
+/-- winner: The participant with the highest bid. -/
+noncomputable def winner : a.I := Classical.choose (exists_max b)
+
+/-- The bid of the winner equals the highest bid. -/
+lemma winner_take_max : b (winner b) = maxb b:= Classical.choose_spec (exists_max b)
+
+/-- Removing a participant `i` from all participants still leaves a non-empty set. -/
+lemma delete_i_nonempty (i:a.I) :Finset.Nonempty (Finset.erase  Finset.univ i ) := by
+  obtain ⟨ i , hi ⟩  := a.hP' i
+  use i
+  simp only [Finset.mem_univ, not_true, Finset.mem_erase, and_true]
+  rw [ne_comm]
+  exact hi
+
+/-- `B i` is the maximal bid of all participants but `i`. -/
+noncomputable def B (i: a.I) : ℝ  := Finset.sup' (Finset.erase Finset.univ i)
+(delete_i_nonempty i) b
+
+/--The second highest bid: the highest bid excluding the winner’s bid.-/
+noncomputable def secondprice : ℝ  := B b (winner b)
+
+/-- Defines the utility of participant `i`,
+which is their valuation minus the second highest bid if `i` is the winner, otherwise, it's 0. -/
+noncomputable def utility  (i : a.I) : ℝ := if i = winner b then a.v i - secondprice b else 0
+
+/-! ### Proofs and Lemmas -/
+variable {i: a.I}
+/-- If `i` is the winner, then their utility is their valuation minus the second highest bid. -/
+lemma utility_winner  (H: i = winner b) : utility b i = a.v i - secondprice b:= by
+  rw [utility]; simp only [ite_true, H]
+
+/-- If `i` is not the winner, then their utility is 0. -/
+lemma utility_loser (i: a.I) (H : i≠ winner b) : utility b i = 0 := by
+  rw [utility]; simp only [ite_false, H]
+
+/-- A strategy is dominant if bidding `bi` ensures that
+`i`'s utility is maximized relative to any other bids `b'` where `b i = bi`. -/
+def dominant (i : a.I) (bi : ℝ) : Prop :=
+  ∀ b b': a.I → ℝ , (b i = bi) → (∀ j: a.I, j≠ i→ b j = b' j)
+   →  utility  b i ≥ utility b' i
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma gt_wins (i : a.I) (H: ∀ j , i ≠ j →  b i > b j) : i = winner b := by
+   have HH : ∀ j, i = j ↔  b j = maxb b:= by
+      have imax : b i = maxb b := by
+         have H1 : b i ≤  maxb b := by
```

Comment:
A general style comment: mathlib uses a three-space indent not a two-space indent.

Reply:
You mean, the other way around: mathlib uses two spaces. Please change the PR from three spaces to two.

Reply:
Thank you both for your feedback. I understand there seems to be some confusion regarding the indentation style. Based on the guidelines and @eric-wieser comment, I will update the PR full context to use a 3-space indent. If there are any further clarifications or corrections needed, please let me know.

Reply:
Oh no - mathlib indentation is two spaces! (Or four spaces if you indent twice, six space if you indent thrice, etc.)
The [style guide](https://leanprover-community.github.io/contribute/style.html) is pretty clear, in my opinion. If you find a place which differs, please let us know so it can be fixed!

Reply:
Yes sorry, the above is a typo!